engines/buried/avi_frames.cpp

```diff
@@ -0,0 +1,199 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * Additional copyright for this file:
+ * Copyright (C) 1995 Presto Studios, Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "buried/avi_frames.h"
+#include "buried/buried.h"
+#include "buried/graphics.h"
+
+#include "common/system.h"
+#include "graphics/surface.h"
+#include "video/avi_decoder.h"
+
+namespace Buried {
+
+AVIFrames::AVIFrames(const Common::String &fileName, uint cachedFrames) {
+	_maxCachedFrames = 0;
+	_video = 0;
+	_cacheEnabled = false;
+	_lastFrame = 0;
+	_lastFrameIndex = -1;
+	_tempFrame = 0;
+
+	if (!fileName.empty())
+		open(fileName, cachedFrames);
+}
+
+AVIFrames::~AVIFrames() {
+	close();
+}
+
+bool AVIFrames::open(const Common::String &fileName, uint cachedFrames) {
+	if (fileName.empty())
+		return false;
+
+	if (_fileName == fileName)
+		return true;
+
+	close();
+
+	_video = new Video::AVIDecoder();
+
+	if (!_video->loadFile(fileName)) {
+		close();
+		return false;
+	}
+
+	_fileName = fileName;
+
+	// Put us into dither mode, for the 8bpp version
+	BuriedEngine *vm = (BuriedEngine *)g_engine;
+	if (!vm->isTrueColor())
+		_video->setDitheringPalette(vm->_gfx->getDefaultPalette());
+
+	if (cachedFrames == 0) {
+		_cacheEnabled = false;
+	} else {
+		_maxCachedFrames = cachedFrames;
+		_cacheEnabled = true;
+	}
+
+	_lastFrameIndex = -1;
+	return true;
+}
+
+void AVIFrames::close() {
+	delete _video;
+	_video = 0;
+
+	_fileName.clear();
+
+	flushFrameCache();
+
+	_lastFrameIndex = -1;
+	_lastFrame = 0;
+
+	if (_tempFrame) {
+		_tempFrame->free();
+		delete _tempFrame;
+		_tempFrame = 0;
+	}
+}
+
+const Graphics::Surface *AVIFrames::getFrame(int frameIndex) {
+	if (!_video)
+		return 0;
```

Comment:
Agreed. Will amend these cases and search for other similar cases and add as new commits.